pmacApp/src/pmacController.cpp

```diff
@@ -4532,30 +4589,87 @@ asynStatus pmacController::tScanBuildProfileArray(double *positions, double *vel
     for (index = 0; index < numPoints; index++) {
       positions[index] = (eguProfilePositions_[axis][index] - offset) / resolution;
     }
-    // Iterates over points, applying resolution to velocities or calculating from velocity mode
-    for (index = 0; index < numPoints; index++) {
-      if(calculateVel == PMAC_TRAJ_VELOCITY_PROVIDED){
+    if(calculateVel == PMAC_TRAJ_VELOCITY_PROVIDED) {
+      // Iterates over points, applying resolution to velocities
+      for (index = 0; index < numPoints; index++) {
         velocities[index] = (eguProfileVelocities_[axis][index]) / resolution;
-      } else if(calculateVel == PMAC_TRAJ_VELOCITY_CALCULATED){
-        status = this->tScanCalculateVelocityArray(positions, velocities, times, index);
-      } else {
+      }
+    } else if(calculateVel == PMAC_TRAJ_VELOCITY_CALCULATED) {
+      // Iterates over points, calculating velocities from velocity modes
+      for (index = 0; index < numPoints; index++) {
+        status = this->tScanCalculateVelocityArray(positions, velocities, times, numPoints, index, csEnum, axis);
+      }
+    } else {
         debug(DEBUG_ERROR, functionName, "Invalid velocity assignment option", calculateVel);
         status = asynError;
-      }
     }
   }
 
   return status;
 }
 
-asynStatus pmacController::tScanCalculateVelocityArray(double *positions, double *velocities, double *times, int index) {
+asynStatus pmacController::tScanCalculateVelocityArray(double *positions, double *velocities, double *times, int numPoints, int index, int csNum, int axis) {
   asynStatus status = asynSuccess;
   static const char *functionName = "tScanCalculateVelocityArray";
 
   double inverse_deltaTime = 0;
   double deltaPos = 0.0;
+  double previousPos = 0.0;
+  double previousVel = 0.0;
+
+  int prevBuffLen = 0;
+  int prevBuffLastVelMode = -1;
+  double prevBuffPosition = 0;
+  double prevBuffVelocity = 0;
+  int prevBuffTime = 0;
+
+  char command[PMAC_MAXBUF_];
+  char reply[PMAC_MAXBUF_];
+
+  // Check if there is a velocity calculation pending from previous buffer
+  if((1 << axis & tScanPendingPoint_) > 0) {
+    prevBuffLen=pTrajectory_->getNoOfValidPoints()-1;
+    // Not first buffer built
+    if(prevBuffLen > 1) {
+      status=pTrajectory_->getVelocityMode(prevBuffLen,&prevBuffLastVelMode);
+      if (status != asynSuccess) {
+        debug(DEBUG_ERROR, functionName, "Failed to get previous Velocity Mode");
+      }
+      else if(prevBuffLastVelMode == 0) {
+         // Prev->Next
+        prevBuffPosition = tScanPrevBufferPositions[axis][0];
+        prevBuffTime = tScanPrevBufferTime;
+
+        if(times[0] <=  0 && prevBuffTime <= 0) {
+          debugf(DEBUG_ERROR, functionName, "Invalid times (%d and %d) - index %d and pending Time from previous buffer",
+                 times[0],prevBuffTime, index, (index+1));
+          status = asynError;
+        }
+
+        inverse_deltaTime = 1000000 / (prevBuffTime+times[0]);
```

Comment:
1000000 is used as a magic number twice in this section.  What does it represent?